### DISCUSSION TITLE
Subtracting interior, non axis-aligned boxes

### DISCUSSION THREAD

**Discussion by starseeker** [elalish/manifold#614] [Q&A] [answered]
I'm not sure if this is an issue, something about how I'm processing, or simply an inevitability of floating point and the boolean algorithms themselves, but I've encountered a surprising result from a boolean operation...

The input is an interior box subtracted from an exterior box (neither is axis aligned, but they appear to be parallel relative to each other:
![image](https://github.com/elalish/manifold/assets/238416/566fdb25-59bf-4410-817c-f93f138a5da8)

![image](https://github.com/elalish/manifold/assets/238416/16008475-65e9-4f67-ae46-0077e29190e5)

The "ground truth" expectation for the evaluated form of this shape (per our raytracing of the CSG boolean tree) is a frame:

![image](https://github.com/elalish/manifold/assets/238416/a2a08e19-6610-40c2-907d-d54023e50e74)

The result I am getting when I convert the two shapes to meshes and subtract them is:

![image](https://github.com/elalish/manifold/assets/238416/0d9dc0bf-4b27-4eae-ae3f-2a9c673b2cb6)

![image](https://github.com/elalish/manifold/assets/238416/71aa55a5-005b-4200-82b3-1f8064853cdc)

I've pulled .glb files for the two inputs and the evaluated subtraction, in case they are useful.
https://brlcad.org/~starseeker/s.wind6.glb
https://brlcad.org/~starseeker/s.wind6.i.glb
https://brlcad.org/~starseeker/s.wind6-s.wind6.i.glb

If this is an expected phenomena I apologize in advance for the noise, but since for our use case it registers visually as a "wrong" result (albeit manifold) I wanted to check and see if there were any good options...

**Comment by pca006132** [comment]
Are the height of the two objects exactly the same? This can be caused by floating point errors and the usual advice is to increase the height of the inner box slightly. Not sure if there are better ways.

**Comment by starseeker** [comment]
OK.  I was afraid that might be the case.

**Comment by elalish** [comment, marked as answer]
Indeed, this is expected behavior. What you have here we call "marginal geometry"; from our [wiki](https://github.com/elalish/manifold/wiki/Manifold-Library#overlaps):
>By this definition, the inputs and results are considered approximate, bounded by their respective precisions which are tracked internally. Therefore, any geometric differences within this precision are considered "in the noise" and are not considered to affect the geometric validity of the result.

If you do the Boolean first and then the rotations, you'll probably get the expected result, as then the coords are exactly identical and our symbolic perturbation does the right thing. But there's simply no way to get exactly coplanar results when things are rotated, since you've introduced rounding errors.

**Comment by OLDMAN75914091** [comment]
This is a rather important test case -  can often happen in real life. 

![image](https://github.com/user-attachments/assets/869b9274-d2bb-43d2-84fd-8428df5968cb)


Please answer the 3 questions asked in the above image.

**Reply by elalish** [reply]
Hard to say - I mostly just see z-fighting which likely implies thin surfaces - manifoldness has to be checked topologically, not by eye. If you're concerned we don't create manifold output, please create a `TEST` case that demonstrates the failure and push it as a PR - we'll take fixing that very seriously. But please don't just *guess* that our algorithms will fail. :)

**Reply by OLDMAN75914091** [reply]
No, not at all suggesting failure......

For (3) - possibly z-fighting - so there could be overlapping triangles there- which is a good sign - as it might explain how that area is valid topologically. 

I had a closer look at the pic - can probably now explain how the overall mesh is topologically valid despite these gaps (1) and (2) - possibly overlapping  triangles in the region bounding these gaps -  more good news for me

Are these explanations plausible?

**Reply by starseeker** [reply]
It tests as topologically valid per the Manifold library - we should fail immediately if we somehow get something out that isn't manifold according to Manifold's tests.  For 1 and 2 they are empty spaces, but not topological gaps in the mesh - 3 (and the other interior triangles) are super thin pairings, as far as I can tell - it's basically the Boolean logic guaranteeing robustness and manifoldness despite the floating point quirks.  The modeling intent in this particular case is "exactly parallel" but Manifold can't recognize that, and that intent can't really be safely assumed in all such cases. 

What I try to do in BRL-CAD is post-process a bit and use our solid raytracing capability to recognize when we have those super-close cases.  I strip those triangles out (temporarily creating a non-manifold) and then use Geogram's hole filling routines to try and close the mesh.  In the above case that works and produces a manifold, but it doesn't always fly - if the cull-and-fill routine doesn't produce valid output I keep and return the original manifold, but there's no denying those thin triangles are problematic in some situations and something for users to be aware of.

The "real" fix in our case is to perturb our original CSG geometry a little to get it out of the coplanar configuration, but that's not as easy as one might think to do automatically.  I'm pondering some ideas, but even if I do come up with something it's not likely to be applicable outside of our own CSG tree hierarchies.

I mentioned that example so you wouldn't be startled if you saw something similar - the first instinctual response when one sees something like that is "auugh!  bug!" but that's not (at least, not in any case I've seen in our inputs) what's going on.

**Reply by OLDMAN75914091** [reply]
crystal clear -- thank you

**Reply by OLDMAN75914091** [reply]
"    The "real" fix in our case is to perturb our original CSG geometry a little to get it out of the coplanar configuration, but that's not as easy as one might think to do automatically. I'm pondering some ideas, but even if I do come up with something it's not likely to be applicable outside of our own CSG tree hierarchies.   "

Agreed that Manifold has produced a topologically manifold result in your example. Is that "extra" geometry in the Boolean result arising due to symbolic perturbation? And if so, then isn't it possible that your attempts at vertex perturbation in your model upstream to Manifold might also produce such things?

OLDMAN

**Reply by starseeker** [reply]
If done naively, that's possible (although statistically less likely if the perturbations are random). That's why I say the CSG Boolean tree is likely to be important - my current thought is I'll have to investigate the models, identify coplanar or near-coplanar faces ahead of time, and selectively perturb in or out along the face normals based on the specific Boolean operation + surface configuration at play in each case. If I perturb coplanar faces rather than vertices directly (readily possible with certain of our implicit primitives) I should be able to avoid most cases where I'd be perturbing into a problem case. It would probably be hard to come up with an algorithm that could be formally proved to be reliable, but if we can come up with something that is practical for the common subset of real world cases we see in practice (sort of like Manifold's epsilon polygon triangulation) then it would still be much better than having to manually adjust everything.

**Reply by OLDMAN75914091** [reply]
I have a project where I am also contending with such things. And for me all such interventions need to be totally automatic/programmatic - not practical to manually grab and perturb scene objects.

What I have attempted is this: Before every pairwise Boolean, choose any one of two parent objects and ever-so-slightly rotate it by a very small angle about a certain 3D vector passing through its centroid. I choose this 3D vector to be something irregular - meaning not the global X or Y or Z vector directions and the like. And after the Boolean I undo this object rotation to keep the overall scene geometry undisturbed.

I do this for ALL Booleaned object-pairs, not just for objects with planar faces since the problem can arise in other cases too, I think. 

As you stated, this intervention works sometimes but not always. I'm still not entirely clear on why it sometimes fails to produce a manifold resulting object. I see two possible reasons: (a) bugs in this particular Boolean implementation code, (b) inappropriate choice of perturbation values (3D-vector and rotation-angle) for that specific object pair.

And this is exactly what prompted me to search for a better solution - leading to my discovery of Manifold.

**Reply by starseeker** [reply]
Hmm.  Using the above rectangle-within-rectangle case as an example, my naive expectation would be that such a rotation (depending on which vector you chose) would produce something like what you actually see in the image above? (Assuming it didn't just fail outright to produce a manifold output...)  Wouldn't you be doing basically what the floating point fuzz and/or rounding errors did inadvertently with my example?

**Reply by elalish** [reply]
Yes, those perturbations sometimes fix an unreliable computational geometry algorithm, but they can't fix everything. If your coplanar cases tend to be axis-aligned, Manifold's symbolic perturbation is designed carefully to avoid generating thin webs and other undesirable geometry. If you can avoid introducing rounding errors by things like inexact rotations, you'll get the best results. Our own `Rotate()` function is designed to be exact for any multiple of 90 degrees (part of why we use degrees instead of radians).

**Reply by OLDMAN75914091** [reply]
That's correct, starseeker.  And yes, it did fail outright to produce a manifold output since, as I now surmise, the faceted Boolean package I am using there is imperfect. 

Also, this expectation of co-planar cases being axis-aligned is really not reasonable for real-life cases.

Conclusion: I should give Manifold a try :)
